streamz/tests/test_sinks.py

```diff
@@ -0,0 +1,59 @@
+from time import sleep
+
+import pytest
+from streamz import Stream
+from streamz.sinks import _global_sinks
+from streamz.utils_test import tmpfile
+
+
+def test_sink_with_args_and_kwargs():
+    L = dict()
+
+    def mycustomsink(elem, key, prefix=""):
+        key = prefix + key
+        if key not in L:
+            L[key] = list()
+        L[key].append(elem)
+
+    s = Stream()
+    s.sink(mycustomsink, "cat", "super")
+
+    s.emit(1)
+    s.emit(2)
+    assert L['supercat'] == [1, 2]
+
+
+def test_sink_to_textfile_fp():
+    source = Stream()
+    with tmpfile() as filename, open(filename, "w", buffering=1) as fp:
+        source.map(str).sink_to_textfile(fp)
+        source.emit(0)
+        source.emit(1)
+
+        sleep(0.01)
+
+        assert open(filename, "r").read() == "0\n1\n"
+
+
+def test_sink_to_textfile_named():
+    source = Stream()
+    with tmpfile() as filename:
+        source.map(str).sink_to_textfile(filename)
+        source.emit(0)
+        source.emit(1)
+
+        sleep(0.01)
+
+        assert open(filename, "r").read() == "0\n1\n"
+
+
+def test_sink_to_textfile_closes():
+    source = Stream()
+    with tmpfile() as filename:
+        sink = source.sink_to_textfile(filename)
+        fp = sink._fp
+        _global_sinks.remove(sink)
```

Comment:
Is there no public way to do this? It should be possible to destruct sinks (and their whole graph).

Reply:
I thought about this. It's tricky to figure out what exactly the user wants to destroy, especially with complex graphs. You don't want to keep a reference to the sink in a variable, so this should probably be a method of `Stream` that can be called on any node and then propagated through the graph. But if there are many branches, you don't know where it should be propagated. Maybe something like `.destroy_downstream()`?

Reply:
This exists, I think in https://github.com/python-streamz/streamz/blob/master/streamz/core.py#L250 and https://github.com/python-streamz/streamz/blob/master/streamz/core.py#L506

Reply:
I should note that the existing `destroy` should use the `_remove_upstream` methods since that handles things properly.

Reply:
In terms of semantics, `.destroy` sounds like something is permanently deleted, while in reality it works more like `disconnect_from_upstreams`. If there's a sink at the end, nothing is destroyed. Maybe it should call `.destroy` on all downstreams, and then for sinks we should override `destroy` to also delete itself from `_global_sinks`?

Reply:
An explicit destroy on a sink should remove it from the global sinks I think, yes. That can then be what gets called here in the test.

Reply:
> I should note that the existing destroy should use the _remove_upstream methods since that handles things properly.

I tried doing this and `test_destroy` in core fails, because `_remove_upstream` replaces an upstream with `[None]` if there's only one upstream:

https://github.com/python-streamz/streamz/blob/a6e91114d6d3a9e2965856ffaaf3e56b6e9d3dfa/streamz/core.py#L257-L258

Why is a single upstream a special case? I'd expect just `self.upstreams.remove(upstream)` here.

Reply:
I think it was so you had the option to add upstreams in case you needed to, see the associate addition method above

Reply:
But the condition in `_add_upstream` will never be satisfied:

https://github.com/python-streamz/streamz/blob/a6e91114d6d3a9e2965856ffaaf3e56b6e9d3dfa/streamz/core.py#L238-L244

if you called `_remove_upstream` before, `self.upstreams` is `[[None]]`, not `[None]`. And why not just use `append` for all cases?

Reply:
Also `[None]` is the state of a stream with no upstream or upstreams. https://github.com/python-streamz/streamz/blob/a6e91114d6d3a9e2965856ffaaf3e56b6e9d3dfa/streamz/core.py#L166

Reply:
The `[[None]]` may be a mistake.

Reply:
> Also [None] is the state of a stream with no upstream or upstreams.

Why not just an empty list? It seems there's a lot of code just to support this `else` branch in `Stream.__init__`:

https://github.com/python-streamz/streamz/blob/a6e91114d6d3a9e2965856ffaaf3e56b6e9d3dfa/streamz/core.py#L168-L171

> The [[None]] may be a mistake.

I'll try to make in `[None]` and rerun tests.